src/bun.js/bindings/JSMockFunction.cpp

```diff
@@ -1192,35 +1217,195 @@ JSC_DEFINE_HOST_FUNCTION(jsMockFunctionMockRejectedValue, (JSC::JSGlobalObject *
+{
+    auto& vm = jsGlobalObject->vm();
+    auto count = callframe->argumentCount();
+    auto ctx = jsDynamicCast<MockWithImplementationCleanupData*>(callframe->argument(1));
+    if (!ctx) {
+        return JSValue::encode(jsUndefined());
+    }
+
+    auto fn = jsDynamicCast<JSMockFunction*>(ctx->internalField(0).get());
+    fn->implementation.set(vm, fn, ctx->internalField(1).get());
+    fn->tail.set(vm, fn, ctx->internalField(2).get());
+    fn->fallbackImplmentation.set(vm, fn, ctx->internalField(3).get());
+
+    return JSValue::encode(jsUndefined());
+}
+JSC_DEFINE_HOST_FUNCTION(jsMockFunctionWithImplementation, (JSC::JSGlobalObject * jsGlobalObject, JSC::CallFrame* callframe))
 {
+    Zig::GlobalObject* globalObject = jsCast<Zig::GlobalObject*>(jsGlobalObject);
+
     JSMockFunction* thisObject = jsDynamicCast<JSMockFunction*>(callframe->thisValue().toThis(globalObject, JSC::ECMAMode::strict()));
     auto& vm = globalObject->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
     if (UNLIKELY(!thisObject)) {
         throwTypeError(globalObject, scope, "Expected Mock"_s);
+        return {};
+    }
+
+    JSValue tempImplValue = callframe->argument(0);
+    JSValue callback = callframe->argument(1);
+    JSC::CallData callData = JSC::getCallData(callback);
+    if (UNLIKELY(callData.type == JSC::CallData::Type::None)) {
+        throwTypeError(globalObject, scope, "Expected mock implementation to be callable"_s);
+        return {};
     }
 
-    HANDLE_STATIC_CALL;
+    auto lastImpl = thisObject->implementation.get();
+    auto lastTail = thisObject->tail.get();
+    auto lastFallback = thisObject->fallbackImplmentation.get();
 
-    if (callframe->argumentCount() < 1) {
-        pushImplOnce(thisObject, globalObject, JSMockImplementation::Kind::Promise, JSC::JSPromise::rejectedPromise(globalObject, jsUndefined()));
-    } else {
-        pushImplOnce(thisObject, globalObject, JSMockImplementation::Kind::Promise, JSC::JSPromise::resolvedPromise(globalObject, callframe->argument(0)));
+    JSMockImplementation* impl = JSMockImplementation::create(
+        globalObject,
+        globalObject->mockModule.mockImplementationStructure.getInitializedOnMainThread(globalObject),
+        JSMockImplementation::Kind::Call,
+        tempImplValue,
+        false);
+
+    thisObject->implementation.set(vm, thisObject, impl);
+    thisObject->fallbackImplmentation.clear();
+    thisObject->tail.clear();
+
+    MarkedArgumentBuffer args;
+    NakedPtr<Exception> exception;
+    JSValue returnValue = call(globalObject, callback, callData, jsUndefined(), args, exception);
+
+    if (auto promise = tryJSDynamicCast<JSC::JSPromise*>(returnValue)) {
+        auto capability = JSC::JSPromise::createNewPromiseCapability(globalObject, globalObject->promiseConstructor());
+
+        JSFunction* func = JSC::JSFunction::create(vm, globalObject, 2,
+            String(), jsMockFunctionWithImplementationCleanup, ImplementationVisibility::Public);
```

Comment:
This function can be cached on the JSGlobalObject so that it doesn't need to be re-created each invocation

Reply:
LazyProperty<JSGlobalObject, JSFunction>